File: clojure/10.clj
(defn pack [lst]
  (if (empty? lst)
    [[]]
    (if (= (count lst) 1)
      [lst]
      (let [packtail (pack (rest lst))]
        (if (= (first lst) (first (rest lst)))
          (cons (cons (first lst)
                      (first packtail))
                (rest packtail))
          (cons [(first lst)] packtail))))))

(defn encode [lst]
  (let [grouped_list (pack lst)]
  (map #(list (count %) (first %)) grouped_list)))

File: scala/10.scala
object Solution10 {
  def pack[T](lst: List[T]): List[List[T]] = {
    lst match {
      case List() => List(List())
      case List(h) => List(List(h))
      case x::y::_ => {
        val packtail = pack(lst.tail)
        if (x == y) (x :: packtail.head) :: packtail.tail
        else List(x) :: packtail
      }
    }
  }

  def encode[T](lst: List[T]): List[(Int, T)] = {
    val packed = pack(lst)
    packed.map(lst => (lst.length, lst.head))
  }
}
